### PR TITLE
#5097: Support for dedicated completion queue thread

### PR DESCRIPTION
**Pull request by DrJessop**
This PR enables support for dedicated completion queue thread and events written to completion queue.

Details:
- Dedicated thread for completion queue, constantly stepping through the completion queue when the comp_write ptr updates
- Support for non-blocking reads 
- Finish semantics have simplified. Finish just waits for num_issued_commands to reach 0

### CODE REVIEW

File: tt_metal/impl/dispatch/command_queue.hpp [hunk 1]
```diff
@@ -339,8 +347,64 @@ class Trace {
 
 namespace detail {
 void EnqueueRestart(CommandQueue& cq);
+
+struct IssuedReadData {
```
Comment (pgkeller) [inline tt_metal/impl/dispatch/command_queue.hpp:351]: can you add a 2 line comment about what this is used for?

File: tt_metal/impl/dispatch/command_queue.hpp [hunk 2]
```diff
@@ -339,8 +347,64 @@ class Trace {
...
+        this->num_pages_read = num_pages_read;
+        this->cur_host_page_id = cur_host_page_id;
+    }
+};
+
+inline std::mutex issued_read_mutex;
+inline std::mutex completion_wrap_mutex;
+
+template <typename T, std::mutex& my_mutex>
+class thread_safe_map {
```
Comment (pgkeller) [inline tt_metal/impl/dispatch/command_queue.hpp:372]: same w/ regard to comment

**Review by pgkeller — COMMENTED** [review]
I don't follow this particularly well, requested some comments.
Is the re_init temporary or permanent?  Not excited about pushing temporary changes up through the API (eg, to tt_eager).

**Comment by DrJessop** [comment]
@pgkeller "Don't follow this particularly well" could you expand on this?

"Is the re-init temporary or permanent" definitely temporary. The reason this was done is because the GetCommandQueue API was originally reconstructing a CQ each time it was being called. That means on subsequent calls, the CommandQueue destructor is called. Unfortunately, I assert now that we don't have any commands in flight by the time we call the destructor.  I essentially now want to ensure that we only make a command queue when we initialize a device, which is of course temporary. A better solution would be for the device to own the command queue.

If you can think of a cleaner way to achieve this with the least possible code changes and without changing the API, let me know. The only other way I can think of is with a global, which can be marked as temporary and won't change the API.

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 3]
```diff
@@ -953,66 +980,69 @@ void EnqueueProgramCommand::process() {
...
-    if (this->wrap_region == DeviceCommand::WrapRegion::COMPLETION) {
-        // Wrap the read pointers for completion queue because device will start writing data at head of completion queue and there are no more reads to be done at current completion queue write pointer
-        // If we don't wrap the read then the subsequent read buffer command may attempt to read past the total command queue size
-        // because the read buffer command will see updated write pointer to compute num pages to read but the local read pointer is pointing to tail of completion queue
-        this->manager.wrap_completion_queue_rd_ptr(this->command_queue_id);
-        this->manager.issue_queue_push_back(wrap_packet_size_bytes, LAZY_COMMAND_QUEUE_MODE, this->command_queue_id);
-    } else {
-        this->manager.wrap_issue_queue_wr_ptr(this->command_queue_id);
-    }
+    this->manager.wrap_next_completion_queue_wr_ptr(this->command_queue_id);
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue.cpp:1037]: so host is keeping track of additional completion queue pointer? I don't know if I understand what "wrap next pointer" signifies

File: tt_metal/detail/tt_metal.hpp [hunk 4]
```diff
@@ -267,7 +267,7 @@ namespace tt::tt_metal{
             return true;
         }
 
-        inline CommandQueue &GetCommandQueue(Device *device)
+        inline CommandQueue &GetCommandQueue(Device *device, bool init_cq)
```
[the remark this answers is not in the document]
Comment (TT-billteng) [reply tt_metal/detail/tt_metal.hpp:270]: would also minimize how many conflicts I need to go through when I merge :D

File: tt_metal/impl/dispatch/command_queue_interface.hpp [hunk 5]
```diff
@@ -302,13 +360,19 @@ class SystemMemoryManager {
...
+    }
+
     void completion_queue_pop_front(uint32_t data_read_B, const uint8_t cq_id) {
         uint32_t data_read_16B = align(data_read_B, 32) >> 4;
 
         SystemMemoryCQInterface& cq_interface = this->cq_interfaces[cq_id];
         cq_interface.completion_fifo_rd_ptr += data_read_16B;
         if (cq_interface.completion_fifo_rd_ptr >= cq_interface.completion_fifo_limit) {
-            cq_interface.completion_fifo_rd_ptr = cq_interface.command_issue_region_size >> 4;
+            cq_interface.completion_fifo_rd_ptr = cq_interface.issue_fifo_limit;
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue_interface.hpp:375]: should the names here also have 16B suffix?

File: tt_metal/impl/dispatch/command_queue.cpp (hunk 3 above)
Comment (DrJessop) [reply tt_metal/impl/dispatch/command_queue.cpp:1037]: hopefully was clearer after our code review meeting.

**Review by TT-billteng — APPROVED** [review]

**Pull request merged**